File: 813/c.cpp
#include <bits/stdc++.h>
using namespace std;


vector<int> nums, bien, saw, uniques, seen;
int solve() {
    int N; cin >> N;
    nums = vector<int>(N);
    uniques = vector<int>(N);
    bien = vector<int>(int(1e5) + 5, 0);
    seen = vector<int>(int(1e5) + 5, 0);
    set<int> uni;
    for (int i = 0; i < N; i++) {
        int d; cin >> d;
        uni.emplace(d);
        uniques[i] = uni.size();
        nums[i] = d;
    }

    int jzong{}, zong{};

    int cp{-1};
    for (int i = N - 1; i > 0; i--) {
        if (nums[i] >= nums[i-1]) {
            seen[nums[i]] = i;
        } else {
            cp = i - 1;
            break;
        }
    }

    if (cp != -1) {
        while (cp >= 0) {
            if (seen[nums[cp]])
                jzong = max(jzong, uniques[seen[nums[cp]]]);
            else {
                if (!bien[nums[cp]]) {
                    bien[nums[cp]] = 1;
                    zong++;
                }
            }
            cp--;
        }
        return max(zong, jzong);
    } else return 0;
}

int main() {
    int t; cin >> t;
    while (t--) cout << solve() << '\n';
}

File: 813/c.py
for _ in range(int(input())):
    N = int(input())
    nums = list(map(int, input().split()))
    bien = [0 for _ in range(10**5 + 5)]
    seen = [0 for _ in range(10**5 + 5)]
    uniques = [0 for _ in range(N + 1)]
    jzong = 0
    zong = 0
    diffs = len(set(nums))

    cp = None
    for i in range(N-1, 0, -1):
        if nums[i] >= nums[i-1]:
            if not seen[nums[i]]: seen[nums[i]] = i
            else: diffs-= 1
            uniques[i] = diffs
                
        else: 
            cp = i - 1
            break
    if cp:
        while cp >= 0:
            if seen[nums[cp]]:
                jzong = max(jzong, uniques[seen[nums[cp]]])
            else:
                if not bien[nums[cp]]:
                    bien[nums[cp]] = 1
                    zong+= 1
            cp-= 1
        print(max(zong, jzong))
    else:
        print(0)